# iSeeGC

iSeeGC is a powerfull bioinformatic pipeline that allows identifying recurrent gene conversion across lineages. It is written by [Sishuo Wang](https://orcid.org/0000-0002-7220-7305) (sishuowang{at}hotmail.com) from University of British Columbia.

# Installation
iSeeGC is run in Linux environment.

Many scripts included in this product require packages from
* [BioRuby](http://bioruby.org)
* [BioPerl](http://bioperl.org)

This product also employs several other computational tools. Please ensure that you have them installed.
* [OrthoMCL](http://orthomcl.org/orthomcl/) 
* [MAFFT](http://mafft.cbrc.jp/alignment/software/)
* [FastTree](http://darlinglab.org/blog/2015/03/23/not-so-fast-fasttree.html)
* [RAxML](https://sco.h-its.org/exelixis/software.html)


Please type **"bash check_requirements.sh"** to see whether required software and libraries have been installed.

# Usage
Type iSeeGC.sh -h to the help message.

# Caveats
![alt text](https://upload.wikimedia.org/wikipedia/commons/thumb/2/2f/Google_2015_logo.svg/220px-Google_2015_logo.svg.png)

# Copyright and Licence:
**No copyright** is required for iSeeGC.

Please cite coppresponding papers for the use of OrthoMCL, MAFFT, FastTree, RAxML, BioRuby and BioPerl.

# Acknowledgements
Thanks for Michael Imelfort and Kranti Konganti for sharing the free script gff2fasta.pl for sequence processing.


